e2e/assets/playground_backend/service/pool/Main.mo

```diff
@@ -107,8 +118,23 @@ shared (creator) actor class Self(opt_params : ?Types.InitParams) = this {
             };
         };
     };
+    func validateOrigin(origin: Logs.Origin) : Bool {
+        if (origin.origin == "") {
+            return false;
+        };
+        for (tag in origin.tags.vals()) {
+            // reject server side tags
+            if (tag == "mode:install" or tag == "mode:reinstall" or tag == "mode:upgrade" or tag == "wasm:profiling" or tag == "wasm:asset") {
+                return false;
+            }
+        };
+        return true;
+    };
 
-    public shared ({ caller }) func getCanisterId(nonce : PoW.Nonce) : async Types.CanisterInfo {
+    public shared ({ caller }) func getCanisterId(nonce : PoW.Nonce, origin : Logs.Origin) : async Types.CanisterInfo {
```

Comment:
Duplicating the whole playground backend here in the e2e/assets sure is a lot of technical debt. Can you think of a better way to maintain this?

Reply:
I agree. We can git clone motoko-playground repo and build, instead of hardcoding the code here. We only need to build the pool canister, and it's okay to keep the pre-built `wasm-utils.wasm` here. Would be good to do as a follow up PR.